README.md
[![CI](https://github.com/cesbit/qpack/workflows/CI/badge.svg)](https://github.com/cesbit/qpack/actions)
[![Release Version](https://img.shields.io/github/release/cesbit/qpack)](https://github.com/cesbit/qpack/releases)

QPack
=====

QPack is a fast and efficient serialization format like MessagePack.
One key difference is flexible map and array support which allows
to write directly to a qpack buffer without the need to know
the size for the map or array beforehand.


Installation
------------

From PyPI (recommend)

```
pip install qpack
```

From source code

```
python setup.py install
```

Pack
----

`qpack.packb(object)`

Unpack
----

Unpack serialized data. When decode is left None, each string
will be returned as bytes.

`qpack.unpackb(qp, decode=None, ignore_decode_errors=False, use_tuples=False)`


Example
-------

```python
import qpack

# define some test data
data = {'name': 'Iris', 'age': 9}

# serialize into qpack format
qp = qpack.packb(data)

# unpack the serialized data
unpacked = qpack.unpackb(qp, decode='utf-8')

# left see what we've got...
print(unpacked)  # {'name': 'Iris', 'age': 3}
```



ChangeLog
2022.09.28, Version 0.0.21

  * Use build-in function.

2022.09.28, Version 0.0.20

  * Added `use_tuples` unpack argument (pr #5).
  * Changed company name to Cesbit.

2020.11.11, Version 0.0.19

  * Use String internal for keys in a map (pr #3).

2019.09.11, Version 0.0.18

  * Added `ignore_decode_errors` keyword argument to unpackb().

2017.08.16, Version 0.0.17

  * Fixed C99 compile errors and warnings.

2017.08.14, Version 0.0.16

  * raise TypeError when trying to pack an unsupported type.
    (issue #2, reported by @pumelo)

2017.03.08, Version 0.0.15 (BETA)

  * Fixed unpacking negative integers using Python2.

  * Replaced adding double value with a memcpy.

2017.03.08, Version 0.0.14 (BETA)

  * Fixed checking bytes size.

2017.02.07, Version 0.0.13 (BETA)

  * Fixed bug in unpacking maps.

2016.10.19, Version 0.0.12 (BETA)

  * Fixed Python2 bugs in fallback.py
  * Strict C99
  * Fixed minor bugs for compiling on Windows platform

2016.10.18, Version 0.0.11 (BETA)

  * Removed fallback print line
  * Removed inttypes.h dependency
  * No declarations in for-loop

2016.10.17, Version 0.0.10 (BETA)

  * Removed p3c dependency

2016.10.14, Version 0.0.9 (BETA)

  * Fixed unicode bug (C-module Python2 only)

2016.10.13, Version 0.0.8 (BETA)

  * Fixed compile bug (C-module Python2)

2016.10.13, Version 0.0.7 (BETA)

  * Fixed Latin-1 decoding (C-module Python2)

2016.10.13, Version 0.0.6 (BETA)

  * Added C module for Python 2 support

2016.10.03, Version 0.0.5 (BETA)

  * Support for Byte-Array in unpack() method.
  * Fixed bug in installing this package using pip.
  * Added C module (Only Python3 support)

2016.10.02, Version 0.0.2 (BETA)

  * Added support for hooks.
  * Fixed Python 2 Compatibility bug.


qpackc/test.py
import imp
import qpack as qp
import time
import msgpack
qpack = imp.load_dynamic('qpack', 'qpack-x86_64.cpython-35m-x86_64-linux-gnu.so')

COUNT = 10000

# data = [1, 2, 3, {'b': True}, {'What?': 'Iriske is lief', 'Hoe oud is ze?': 3}]


data = {
    'series float': [
        [1471254705, 0.1],
        [1471254707, 0.0],
        [1471254710, 0.0]],
    'series integer': [
        [1471254705, 5],
        [1471254708, -3],
        [1471254710, -7]],
    'aggr': [
        [1447249033, 531], [1447249337, 0.0], [1447249633, 535], [1447249937, 0.0],
        [1447250249, 532], [1447250549, 0.0], [1447250868, 530], [1447251168, 0.0],
        [1447251449, 54], [1447251749, 0.0], [1447252049, 0.0], [1447252349, 0.0],
        [1447252649, True], [1447252968, 0.0], [1447253244, 0.0], [1447253549, 0.0],
        [1447253849, 534], [1447254149, 0.0], [1447254449, 0.0], [1447254748, 0.0]
    ]
}

print(data)
import sys

a = msgpack.packb(data)
b = qp.packb(data)
c = qpack._packb(data)

print(a)
print(b)
print(c)

print(len(a), sys.getrefcount(a))
print(len(b), sys.getrefcount(b))
print(len(c), sys.getrefcount(c))

start = time.time()
for i in range(COUNT):
    msgpack.packb(data)
print('MsgPack (C-module) Pack Time: {}'.format(time.time() - start))

start = time.time()
for i in range(COUNT):
    qp.packb(data)
print('QPack (Python)     Pack Time: {}'.format(time.time() - start))

start = time.time()
for i in range(COUNT):
    qpack._packb(data)
print('QPack (C-module)   Pack Time: {}'.format(time.time() - start))


print(sys.getrefcount(5), sys.getrefcount(True))
A = msgpack.unpackb(a)
print(sys.getrefcount(5), sys.getrefcount(True))
B = qp.unpackb(b)
print(sys.getrefcount(5), sys.getrefcount(True))
C = qpack._unpackb(c)
print(sys.getrefcount(5), sys.getrefcount(True))

print(A)
print(B)
print(C)



start = time.time()
for i in range(COUNT):
    msgpack.unpackb(a)
print('MsgPack (C-module) UnPack Time: {}'.format(time.time() - start))

start = time.time()
for i in range(COUNT):
    qp.unpackb(b)
print('QPack (Python)     UnPack Time: {}'.format(time.time() - start))

start = time.time()
for i in range(COUNT):
    qpack._unpackb(c)
print('QPack (C-module)   UnPack Time: {}'.format(time.time() - start))



test/qpack_test.py
# -*- coding: utf-8 -*-
import sys
import qpack
from qpack import fallback
import unittest
import pickle

if sys.version_info[0] == 3:
    INT_CONVERT = int
    PYTHON3 = True
else:
    INT_CONVERT = ord
    PYTHON3 = False


class TestQpack(unittest.TestCase):

    CASES = [
        [u" Hi Qpack", [
            140, 239, 163, 159, 32, 72, 105, 32, 81, 112, 97, 99, 107]],
        [True, [249]],
        [False, [250]],
        [None, [251]],
        [-1, [64]],
        [-60, [123]],
        [-61, [232, 195]],
        [0, [0]],
        [1, [1]],
        [4, [4]],
        [63, [63]],
        [64, [232, 64]],
        [-1.0, [125]],
        [0.0, [126]],
        [1.0, [127]],
        [-120, [232, 136]],
        [-0xfe, [233, 2, 255]],
        [-0xfedcba, [234, 70, 35, 1, 255]],
        [-0xfedcba9876, [235, 138, 103, 69, 35, 1, 255, 255, 255]],
        [120, [232, 120]],
        [0xfe, [233, 254, 0]],
        [0xfedcba, [234, 186, 220, 254, 0]],
        [0xfedcba9876, [235, 118, 152, 186, 220, 254, 0, 0, 0]],
        [-1.234567, [236, 135, 136, 155, 83, 201, 192, 243, 191]],
        [123.4567, [236, 83, 5, 163, 146, 58, 221, 94, 64]],
        [[0.0, 1.1, 2.2], [
            240, 126, 236, 154, 153, 153, 153, 153, 153, 241, 63,
            236, 154, 153, 153, 153, 153, 153, 1, 64]],
        [[10, 20, 30, 40, 50], [242, 10, 20, 30, 40, 50]],
        [[10, 20, 30, 40, 50, 60], [
            252, 10, 20, 30, 40, 50, 60, 254]],
        [[0, {"Names": ["Iris", "Sasha"]}], [
            239, 0, 244, 133, 78, 97, 109, 101, 115, 239, 132, 73, 114,
            105, 115, 133, 83, 97, 115, 104, 97]]]

    def _pack(self, packb):
        for inp, want in self.CASES:
            out = packb(inp)
            self.assertEqual([INT_CONVERT(i) for i in out], want)

    def _unpack(self, unpackb):
        for inp, want in self.CASES:
            out = unpackb(qpack.packb(inp), decode='utf8')
            self.assertEqual(out, inp)

    def test_packb(self):
        self.assertEqual(
            qpack.packb.__doc__,
            'Serialize a Python object to QPack format.')
        self._pack(qpack.packb)

    def test_fallback_packb(self):
        self.assertEqual(
            fallback.packb.__doc__,
            'Serialize to QPack. (Pure Python implementation)')
        self._pack(fallback.packb)

    def test_unpackb(self):
        self.assertNotIn(
            'Pure Python implementation',
            qpack.unpackb.__doc__)
        self._unpack(qpack.unpackb)

    def test_fallback_unpackb(self):
        self.assertIn(
            'Pure Python implementation',
            fallback.unpackb.__doc__)
        self._unpack(fallback.unpackb)

    def test_packb_unsupported(self):
        with self.assertRaises(TypeError):
            fallback.packb({'module': sys})
        with self.assertRaises(TypeError):
            qpack.packb({'module': sys})

    def test_decode(self):
        if not PYTHON3:
            return
        bindata = pickle.dumps({})
        data = ['normal', bindata]
        packed = qpack.packb(data)
        s, b = qpack.unpackb(packed)
        self.assertEqual(type(s).__name__, 'bytes')
        self.assertEqual(type(b).__name__, 'bytes')

        with self.assertRaises(ValueError):
            s, b = qpack.unpackb(packed, decode='utf8')

        s, b = qpack.unpackb(packed, decode='utf8', ignore_decode_errors=True)
        self.assertEqual(type(s).__name__, 'str')
        self.assertEqual(type(b).__name__, 'bytes')

    def test_fallback_decode(self):
        if not PYTHON3:
            return
        bindata = pickle.dumps({})
        data = ['normal', bindata]
        packed = fallback.packb(data)
        s, b = fallback.unpackb(packed)
        self.assertEqual(type(s).__name__, 'bytes')
        self.assertEqual(type(b).__name__, 'bytes')

        with self.assertRaises(ValueError):
            s, b = fallback.unpackb(packed, decode='utf8')

        s, b = fallback.unpackb(
            packed, decode='utf8', ignore_decode_errors=True)
        self.assertEqual(type(s).__name__, 'str')
        self.assertEqual(type(b).__name__, 'bytes')

    def test_use_tuples(self):
        data = tuple([('a', 'b') for _ in range(20)])
        packed = qpack.packb(data)
        result = qpack.unpackb(packed, decode='utf-8', use_tuples=True)
        self.assertEqual(result, data)

    def test_fallback_use_tuples(self):
        data = tuple([('a', 'b') for _ in range(20)])
        packed = qpack.packb(data)
        result = fallback.unpackb(packed, decode='utf-8', use_tuples=True)
        self.assertEqual(result, data)


if __name__ == '__main__':
    unittest.main()


qpack/qpack.py
'''QPack - (de)serializer

:copyright: 2016, Jeroen van der Heijden (Transceptor Technology)
'''
import sys
import struct

# for being Python2 and Python3 compatible
PY_CONVERT = ord if sys.version_info[0] == 2 else int

SIZE8_T = struct.Struct('<B')
SIZE16_T = struct.Struct('<H')
SIZE32_T = struct.Struct('<I')
SIZE64_T = struct.Struct('<Q')

INT8_T = struct.Struct('<b')
INT16_T = struct.Struct('<h')
INT32_T = struct.Struct('<i')
INT64_T = struct.Struct('<q')

DOUBLE = struct.Struct('<d')

# Fixed integer lengths: b'\x00' - '\x3f'
# Fixed negative integer lengths: b'\x40' - '\x7c'
# Fixed doubles: -1.0 0.0 and 1.0  '\x7d', '\x7e', '\x7f'
QP_DOUBLE_N1 = b'\x7d'
QP_DOUBLE_0 = b'\x7e'
QP_DOUBLE_1 = b'\x7f'

# Fixed str lengths: b'\x80' - '\xe3'  (0 .. 99)

QP_RAW8 = b'\xe4'   # 228
QP_RAW16 = b'\xe5'
QP_RAW32 = b'\xe6'
QP_RAW64 = b'\xe7'

QP_INT8 = b'\xe8'   # 232
QP_INT16 = b'\xe9'
QP_INT32 = b'\xea'
QP_INT64 = b'\xeb'

QP_DOUBLE = b'\xec' # 236 this one is 8 bytes, reserve for 4 bytes

START_ARR = 237
QP_ARRAY0 = b'\xed' # 237
QP_ARRAY1 = b'\xee'
QP_ARRAY2 = b'\xef'
QP_ARRAY3 = b'\xf0'
QP_ARRAY4 = b'\xf1'
QP_ARRAY5 = b'\xf2'

START_MAP = 243
QP_MAP0 = b'\xf3'   # 243
QP_MAP1 = b'\xf4'
QP_MAP2 = b'\xf5'
QP_MAP3 = b'\xf6'
QP_MAP4 = b'\xf7'
QP_MAP5 = b'\xf8'

QP_BOOL_TRUE = b'\xf9' # 249
QP_BOOL_FALSE = b'\xfa'
QP_NULL = b'\xfb'

QP_OPEN_ARRAY, N_OPEN_ARRAY = b'\xfc', 252
QP_OPEN_MAP, N_OPEN_MAP = b'\xfd', 253
QP_CLOSE_ARRAY, N_CLOSE_ARRAY = b'\xfe', 254
QP_CLOSE_MAP, N_CLOSE_MAP = b'\xff', 255

_RAW_MAP = {
    ord(QP_RAW8): SIZE8_T,
    ord(QP_RAW16): SIZE16_T,
    ord(QP_RAW32): SIZE32_T,
    ord(QP_RAW64): SIZE64_T}

_NUMBER_MAP = {
    ord(QP_INT8): INT8_T,
    ord(QP_INT16): INT16_T,
    ord(QP_INT32): INT32_T,
    ord(QP_INT64): INT64_T,
    ord(QP_DOUBLE): DOUBLE}

_SIMPLE_MAP = {
    ord(QP_BOOL_TRUE): True,
    ord(QP_BOOL_FALSE): False,
    ord(QP_NULL): None}


def _pack(obj, container):
    if obj is True:
        container.append(QP_BOOL_TRUE)

    elif obj is False:
        container.append(QP_BOOL_FALSE)

    elif obj is None:
        container.append(QP_NULL)

    elif isinstance(obj, int):
        if 64 > obj >= 0:
            container.append(struct.pack("B", obj))
            return
        if 0 > obj >= -61:
            container.append(struct.pack("B", 63 - obj))
            return

        bit_len = obj.bit_length()
        if bit_len < 8:
            container.append(QP_INT8)
            container.append(INT8_T.pack(obj))
        elif bit_len < 16:
            container.append(QP_INT16)
            container.append(INT16_T.pack(obj))
        elif bit_len < 32:
            container.append(QP_INT32)
            container.append(INT32_T.pack(obj))
        elif bit_len < 64:
            container.append(QP_INT64)
            container.append(INT64_T.pack(obj))
        else:
            raise OverflowError(
                'QuickPack allows up to 64bit signed integers, '
                'got bit length: {}'.format(bit_len))

    elif isinstance(obj, float):
        if obj == 0.0:
            container.append(QP_DOUBLE_0)
        elif obj == 1.0:
            container.append(QP_DOUBLE_1)
        elif obj == -1.0:
            container.append(QP_DOUBLE_N1)
        else:
            container.append(QP_DOUBLE)
            container.append(DOUBLE.pack(obj))

    elif isinstance(obj, str):
        b = obj.encode('utf-8')
        l = len(b)
        if l < 100:
            container.append(struct.pack("B", 128 + l))
        elif l < 256:
            container.append(QP_RAW8)
            container.append(SIZE8_T.pack(l))
        elif l < 65536:
            container.append(QP_RAW16)
            container.append(SIZE16_T.pack(l))
        elif l < 4294967296:
            container.append(QP_RAW32)
            container.append(SIZE32_T.pack(l))
        elif l < 18446744073709551616:
            container.append(QP_RAW64)
            container.append(SIZE64_T.pack(l))
        else:
            raise ValueError(
                'raw string length too large to fit in QuickPack: {}'
                .format(l))
        container.append(b)

    elif isinstance(obj, bytes):
        l = len(obj)
        if l < 100:
            container.append(struct.pack("B", 128 + l))
        elif l < 256:
            container.append(QP_RAW8)
            container.append(SIZE8_T.pack(l))
        elif l < 65536:
            container.append(QP_RAW16)
            container.append(SIZE16_T.pack(l))
        elif l < 4294967296:
            container.append(QP_RAW32)
            container.append(SIZE32_T.pack(l))
        elif l < 18446744073709551616:
            container.append(QP_RAW64)
            container.append(SIZE64_T.pack(l))
        else:
            raise ValueError(
                'raw string length too large to fit in QuickPack: {}'
                .format(l))
        container.append(obj)

    elif isinstance(obj, (list, tuple)):
        l = len(obj)
        if l < 6:
            container.append(SIZE8_T.pack(START_ARR + l))
            for value in obj:
                _pack(value, container)
        else:
            container.append(QP_OPEN_ARRAY)
            for value in obj:
                _pack(value, container)
            container.append(QP_CLOSE_ARRAY)

    elif isinstance(obj, dict):
        l = len(obj)
        if l < 6:
            container.append(SIZE8_T.pack(START_MAP + l))
            for key, value in obj.items():
                _pack(key, container)
                _pack(value, container)
        else:
            container.append(QP_OPEN_MAP)
            for key, value in obj.items():
                _pack(key, container)
                _pack(value, container)
            container.append(QP_CLOSE_MAP)


def _unpack(qp, pos, end, decode=None):
    tp = PY_CONVERT(qp[pos])
    pos += 1
    if tp < 64:
        return pos, tp

    if tp < 125:
        return pos, 63 - tp

    if tp < 0x80:
        return pos, float(tp - 126)

    if tp < 0xe4:
        end_pos = pos + (tp - 128)
        return end_pos, qp[pos:end_pos] if decode is None \
            else qp[pos:end_pos].decode(decode)

    if tp < 0xe8:
        qp_type = _RAW_MAP[tp]
        end_pos = pos + qp_type.size + qp_type.unpack_from(qp, pos)[0]
        pos += qp_type.size
        return end_pos, qp[pos:end_pos] if decode is None \
            else qp[pos:end_pos].decode(decode)

    if tp < 0xed: # double included
        qp_type = _NUMBER_MAP[tp]
        return pos + qp_type.size, qp_type.unpack_from(qp, pos)[0]

    if tp < 0xf3:
        qp_array = []
        for _ in range(tp - 0xed):
            pos, value = _unpack(qp, pos, end, decode)
            qp_array.append(value)
        return pos, qp_array

    if tp < 0xf9:
        qp_map = {}
        for _ in range(tp - 0xf3):
            pos, key = _unpack(qp, pos, end, decode)
            pos, value = _unpack(qp, pos, end, decode)
            qp_map[key] = value
        return pos, qp_map

    if tp < 0xfc:
        return pos, _SIMPLE_MAP[tp]

    if tp == N_OPEN_ARRAY:
        qp_array = []
        while pos < end and qp[pos] != N_CLOSE_ARRAY:
            pos, value = _unpack(qp, pos, end, decode)
            qp_array.append(value)
        return pos + 1, qp_array

    if tp == N_OPEN_MAP:
        qp_map = {}
        while pos < end and qp[pos] != N_CLOSE_MAP:
            pos, key = _unpack(qp, pos, end, decode)
            pos, value = _unpack(qp, pos, end, decode)
            qp_map[key] = value
        return pos + 1, qp_map

    raise ValueError('Error in quickpack at position {}'.format(pos))


def packb(obj):
    container = []
    _pack(obj, container)
    return b''.join(container)


def unpackb(qp, decode=None):
    return _unpack(qp, 0, len(qp), decode=decode)[1]


if __name__ == '__main__':
    pass




qpack/fallback.py
'''QPack - (de)serializer

:copyright: 2022, Cesbit
'''
import sys
import struct

# for being Python2 and Python3 compatible
if sys.version_info[0] == 3:
    PYTHON3 = True
    PY_CONVERT = int
    INT_TYPES = int
    STR = str
    intern = sys.intern

    def dict_items(d):
        return d.items()
else:
    PYTHON3 = False
    PY_CONVERT = ord
    INT_TYPES = (int, long)
    STR = (unicode, str)
    range = xrange

    def dict_items(d):
        return d.iteritems()

SIZE8_T = struct.Struct('<B')
SIZE16_T = struct.Struct('<H')
SIZE32_T = struct.Struct('<I')
SIZE64_T = struct.Struct('<Q')

INT8_T = struct.Struct('<b')
INT16_T = struct.Struct('<h')
INT32_T = struct.Struct('<i')
INT64_T = struct.Struct('<q')

DOUBLE = struct.Struct('<d')

QP_HOOK = b'\x7c'
# Fixed integer lengths: b'\x00' - '\x3f'
# Fixed negative integer lengths: b'\x40' - '\x7c'
# Fixed doubles: -1.0 0.0 and 1.0  '\x7d', '\x7e', '\x7f'
QP_DOUBLE_N1 = b'\x7d'
QP_DOUBLE_0 = b'\x7e'
QP_DOUBLE_1 = b'\x7f'

# Fixed str lengths: b'\x80' - '\xe3'  (0 .. 99)

QP_RAW8 = b'\xe4'   # 228
QP_RAW16 = b'\xe5'
QP_RAW32 = b'\xe6'
QP_RAW64 = b'\xe7'

QP_INT8 = b'\xe8'   # 232
QP_INT16 = b'\xe9'
QP_INT32 = b'\xea'
QP_INT64 = b'\xeb'

QP_DOUBLE = b'\xec'  # 236 this one is 8 bytes, reserve for 4 bytes

START_ARR = 237
QP_ARRAY0 = b'\xed'  # 237
QP_ARRAY1 = b'\xee'
QP_ARRAY2 = b'\xef'
QP_ARRAY3 = b'\xf0'
QP_ARRAY4 = b'\xf1'
QP_ARRAY5 = b'\xf2'

START_MAP = 243
QP_MAP0 = b'\xf3'   # 243
QP_MAP1 = b'\xf4'
QP_MAP2 = b'\xf5'
QP_MAP3 = b'\xf6'
QP_MAP4 = b'\xf7'
QP_MAP5 = b'\xf8'

QP_BOOL_TRUE = b'\xf9'  # 249
QP_BOOL_FALSE = b'\xfa'
QP_NULL = b'\xfb'

QP_OPEN_ARRAY, N_OPEN_ARRAY = b'\xfc', 252
QP_OPEN_MAP, N_OPEN_MAP = b'\xfd', 253
QP_CLOSE_ARRAY, N_CLOSE_ARRAY = b'\xfe', 254
QP_CLOSE_MAP, N_CLOSE_MAP = b'\xff', 255

_RAW_MAP = {
    ord(QP_RAW8): SIZE8_T,
    ord(QP_RAW16): SIZE16_T,
    ord(QP_RAW32): SIZE32_T,
    ord(QP_RAW64): SIZE64_T}

_NUMBER_MAP = {
    ord(QP_INT8): INT8_T,
    ord(QP_INT16): INT16_T,
    ord(QP_INT32): INT32_T,
    ord(QP_INT64): INT64_T,
    ord(QP_DOUBLE): DOUBLE}

_SIMPLE_MAP = {
    ord(QP_BOOL_TRUE): True,
    ord(QP_BOOL_FALSE): False,
    ord(QP_NULL): None}


def _pack(obj, container):
    if obj is True:
        container.append(QP_BOOL_TRUE)

    elif obj is False:
        container.append(QP_BOOL_FALSE)

    elif obj is None:
        container.append(QP_NULL)

    elif isinstance(obj, INT_TYPES):
        if 64 > obj >= 0:
            container.append(struct.pack("B", obj))
            return
        if 0 > obj >= -60:
            container.append(struct.pack("B", 63 - obj))
            return

        bit_len = obj.bit_length()
        if bit_len < 8:
            container.append(QP_INT8)
            container.append(INT8_T.pack(obj))
        elif bit_len < 16:
            container.append(QP_INT16)
            container.append(INT16_T.pack(obj))
        elif bit_len < 32:
            container.append(QP_INT32)
            container.append(INT32_T.pack(obj))
        elif bit_len < 64:
            container.append(QP_INT64)
            container.append(INT64_T.pack(obj))
        else:
            raise OverflowError(
                'qpack allows up to 64bit signed integers, '
                'got bit length: {}'.format(bit_len))

    elif isinstance(obj, float):
        if obj == 0.0:
            container.append(QP_DOUBLE_0)
        elif obj == 1.0:
            container.append(QP_DOUBLE_1)
        elif obj == -1.0:
            container.append(QP_DOUBLE_N1)
        else:
            container.append(QP_DOUBLE)
            container.append(DOUBLE.pack(obj))

    elif isinstance(obj, STR):
        b = obj.encode('utf-8')
        n = len(b)
        if n < 100:
            container.append(struct.pack("B", 128 + n))
        elif n < 0x100:
            container.append(QP_RAW8)
            container.append(SIZE8_T.pack(n))
        elif n < 0x10000:
            container.append(QP_RAW16)
            container.append(SIZE16_T.pack(n))
        elif n < 0x100000000:
            container.append(QP_RAW32)
            container.append(SIZE32_T.pack(n))
        elif n < 0x10000000000000000:
            container.append(QP_RAW64)
            container.append(SIZE64_T.pack(n))
        else:
            raise ValueError(
                'raw string length too large to fit in qpack: {}'
                .format(n))
        container.append(b)

    elif isinstance(obj, bytes):
        n = len(obj)
        if n < 100:
            container.append(struct.pack("B", 128 + n))
        elif n < 0x100:
            container.append(QP_RAW8)
            container.append(SIZE8_T.pack(n))
        elif n < 0x10000:
            container.append(QP_RAW16)
            container.append(SIZE16_T.pack(n))
        elif n < 0x100000000:
            container.append(QP_RAW32)
            container.append(SIZE32_T.pack(n))
        elif n < 0x10000000000000000:
            container.append(QP_RAW64)
            container.append(SIZE64_T.pack(n))
        else:
            raise ValueError(
                'raw string length too large to fit in qpack: {}'
                .format(n))
        container.append(obj)

    elif isinstance(obj, (list, tuple)):
        n = len(obj)
        if n < 6:
            container.append(SIZE8_T.pack(START_ARR + n))
            for value in obj:
                _pack(value, container)
        else:
            container.append(QP_OPEN_ARRAY)
            for value in obj:
                _pack(value, container)
            container.append(QP_CLOSE_ARRAY)

    elif isinstance(obj, dict):
        n = len(obj)
        if n < 6:
            container.append(SIZE8_T.pack(START_MAP + n))
            for key, value in dict_items(obj):
                _pack(key, container)
                _pack(value, container)
        else:
            container.append(QP_OPEN_MAP)
            for key, value in dict_items(obj):
                _pack(key, container)
                _pack(value, container)
            container.append(QP_CLOSE_MAP)

    else:
        raise TypeError(
            'packing type {} is not supported with qpack'.format(type(obj)))


def _decode(qp, pos, end_pos, decode, ignore_decode_errors):
    raw = qp[pos:end_pos]

    if decode is None:
        return raw

    if ignore_decode_errors:
        try:
            raw = raw.decode(decode)
        finally:
            return raw

    return raw.decode(decode)


def _unpack(qp, pos, end, decode, ign_dec_err, use_tpls):
    tp = PY_CONVERT(qp[pos])
    pos += 1
    if tp < 64:
        return pos, tp

    if tp < 124:
        return pos, 63 - tp

    if tp == QP_HOOK:
        # This is reserved for an object hook.
        return pos, 0

    if tp < 0x80:
        return pos, float(tp - 126)

    if tp < 0xe4:
        end_pos = pos + (tp - 128)
        return end_pos, _decode(qp, pos, end_pos, decode, ign_dec_err)

    if tp < 0xe8:
        qp_type = _RAW_MAP[tp]
        end_pos = pos + qp_type.size + qp_type.unpack_from(qp, pos)[0]
        pos += qp_type.size
        return end_pos, _decode(qp, pos, end_pos, decode, ign_dec_err)

    if tp < 0xed:  # double included
        qp_type = _NUMBER_MAP[tp]
        return pos + qp_type.size, qp_type.unpack_from(qp, pos)[0]

    if tp < 0xf3:
        qp_array = []
        for _ in range(tp - 0xed):
            pos, value = _unpack(qp, pos, end, decode, ign_dec_err, use_tpls)
            qp_array.append(value)
        return pos, tuple(qp_array) if use_tpls else qp_array

    if tp < 0xf9:
        qp_map = {}
        for _ in range(tp - 0xf3):
            pos, key = _unpack(qp, pos, end, decode, ign_dec_err, use_tpls)
            pos, value = _unpack(qp, pos, end, decode, ign_dec_err, use_tpls)
            if isinstance(key, str):
                intern(key)
            qp_map[key] = value
        return pos, qp_map

    if tp < 0xfc:
        return pos, _SIMPLE_MAP[tp]

    if tp == N_OPEN_ARRAY:
        qp_array = []
        while pos < end and PY_CONVERT(qp[pos]) != N_CLOSE_ARRAY:
            pos, value = _unpack(qp, pos, end, decode, ign_dec_err, use_tpls)
            qp_array.append(value)
        return pos + 1, tuple(qp_array) if use_tpls else qp_array

    if tp == N_OPEN_MAP:
        qp_map = {}
        while pos < end and PY_CONVERT(qp[pos]) != N_CLOSE_MAP:
            pos, key = _unpack(qp, pos, end, decode, ign_dec_err, use_tpls)
            pos, value = _unpack(qp, pos, end, decode, ign_dec_err, use_tpls)
            if isinstance(key, str):
                intern(key)
            qp_map[key] = value
        return pos + 1, qp_map

    raise ValueError('Error in qpack at position {}'.format(pos))


def packb(obj):
    '''Serialize to QPack. (Pure Python implementation)'''
    container = []
    _pack(obj, container)
    return b''.join(container)


def unpackb(qp, decode=None, ignore_decode_errors=False, use_tuples=False):
    '''De-serialize QPack to Python. (Pure Python implementation)'''
    return _unpack(qp, 0, len(qp), decode, ignore_decode_errors, use_tuples)[1]


if __name__ == '__main__':
    pass
